api/agent/call.go

```diff
@@ -344,14 +350,21 @@ func (c *call) End(ctx context.Context, err error) {
 	// TODO: this should be update, really
 	if err := c.ds.InsertCall(ctx, c.Call); err != nil {
 		common.Logger(ctx).WithError(err).Error("error inserting call into datastore")
+		return err
 	}
 
 	if err := c.ds.InsertLog(ctx, c.AppName, c.ID, c.stderr); err != nil {
 		common.Logger(ctx).WithError(err).Error("error uploading log")
+		return err
 	}
 
 	// NOTE call this after InsertLog or the buffer will get reset
 	c.stderr.Close()
+	err := t.fireAfterCall(ctx, c.Model())
+	if err != nil {
+		return fmt.Errorf("AfterCall: %v", err)
```

Comment:
imo we shouldn't rewrite this error with `fmt.Errorf` so that users can implement `models.APIError` in the returned error and have this returned from function invocations in the http response. we could log it so we know where it's from instead?

Reply:
I think we need to consider using an error wrapping/chaining lib, we lose so much context which make it really hard to debug. eg: https://github.com/hashicorp/errwrap

I just spent an hour trying to find where a JSON parsing error was happening for instance.

Reply:
the downsides are... it makes it really hard to bubble up typed errors, what I mean is probably better shown by example: https://play.golang.org/p/297H5RAolz

i.e. `models.APIError` would need to turn into... something else, or we have to explicitly make our own error wrapper that can implement `models.APIError`, or any other error interfaces we may be using / want to use (hairy, we attempted this on the runner project and you end up having to add tests for errors you expect because it's really brittle -- some errors are not so easy to generate, either)

makes ya miss exceptions (ooh it hurts to say) :(

I agree it's kind of gross / impossible to log at the root of every error site, need something.

Reply:
Well that's why these wrapping libs have things like https://godoc.org/github.com/hashicorp/errwrap#ContainsType so you can check for a type through the chain.